optaplanner-core/src/main/java/org/optaplanner/core/impl/localsearch/decider/multithreaded/MultiThreadedLocalSearchDecider.java

```diff
@@ -0,0 +1,407 @@
+        super.phaseEnded(phaseScope);
+        // Tell the move thread runners to stop
+        DestroyOperation<Solution_> destroyOperation = new DestroyOperation<>();
+        for (int i = 0; i < moveThreadCount; i++) {
+            operationQueue.add(destroyOperation);
+        }
+        // TODO This should probably be in a finally that spawns at least the entire phase, maybe even the entire solve
+        ThreadUtils.shutdownAwaitOrKill(executor, logIndentation, "Multithreaded Local Search");
+        long childThreadsScoreCalculationCount = 0;
+        for (MoveThreadRunner moveThreadRunner : moveThreadRunnerList) {
+            childThreadsScoreCalculationCount += moveThreadRunner.scoreDirector.getCalculationCount();
+        }
+        phaseScope.addChildThreadsScoreCalculationCount(childThreadsScoreCalculationCount);
+        moveThreadRunnerList = null;
+        operationQueue = null;
+        resultQueue = null;
+    }
+
+    private ExecutorService createThreadPoolExecutor() {
+        ThreadPoolExecutor threadPoolExecutor
+                = (ThreadPoolExecutor) Executors.newFixedThreadPool(moveThreadCount, threadFactory);
+        if (threadPoolExecutor.getMaximumPoolSize() < moveThreadCount) {
+            throw new IllegalStateException(
+                    "The threadPoolExecutor's maximumPoolSize (" + threadPoolExecutor.getMaximumPoolSize()
+                    + ") is less than the moveThreadCount (" + moveThreadCount + "), this is unsupported.");
+        }
+        return threadPoolExecutor;
+    }
+
+    @Override
+    public void decideNextStep(LocalSearchStepScope<Solution_> stepScope) {
+        int stepIndex = stepScope.getStepIndex();
+        resultQueue.startNextStep(stepIndex);
+        int selectingMoveIndex = 0;
+        for (Move<Solution_> selectingMove : moveSelector) {
+            operationQueue.add(new MoveEvaluationOperation<>(stepIndex, selectingMoveIndex, selectingMove));
+            selectingMoveIndex++; // Increment required before if check
+            if (selectingMoveIndex < selectedMoveBufferSize) {
+                // Fill the buffer so move evaluation can run freely in parallel
+                // For reproducibility, the selectedMoveBufferSize always need to be selected,
+                // even if some of those moves won't be evaluated or foraged
+                continue;
+            }
+            OrderByMoveIndexBlockingQueue.MoveResult<Solution_> result;
+            try {
+                result = resultQueue.take();
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+                break;
+            }
+            if (stepIndex != result.getStepIndex()) {
+                throw new IllegalStateException("Impossible situation: the solverThread's stepIndex (" + stepIndex
+                        + ") differs from the result's stepIndex (" + result.getStepIndex() + ").");
```

Comment:
Is this just some kind of sanity check or have you seen this happen? Just seems like a very interesting situation and wonder if its necessary.

Reply:
It's a fail fast. More specifically, because it says "impossible situation", it's a fail fast that won't be triggered by a user doing something wrong (like most fail fasts), but by an intrinsic bug in optaplanner-core itself.
Because the if check is extremely cheap, it's ok to do this.